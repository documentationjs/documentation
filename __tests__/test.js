var documentationSchema = require('documentation-schema'),
  validate = require('json-schema'),
  documentation = require('../'),
  outputMarkdown = require('../src/output/markdown.js'),
  outputMarkdownAST = require('../src/output/markdown_ast.js'),
  outputHtml = require('../src/output/html.js'),
  normalize = require('./utils').normalize,
  glob = require('glob'),
  pify = require('pify'),
  path = require('path'),
  fs = require('fs'),
  _ = require('lodash'),
  chdir = require('chdir');

var UPDATE = !!process.env.UPDATE;

function makePOJO(ast) {
  return JSON.parse(JSON.stringify(ast));
}

function readOptionsFromFile(file) {
  var s = fs.readFileSync(file, 'utf-8');
  var lines = s.split(/\n/, 20);
  for (var i = 0; i < lines.length; i++) {
    var m = lines[i].match(/^\/\/\s+Options:\s*(.+)$/);
    if (m) {
      return JSON.parse(m[1]);
    }
  }
  return {};
}

if (fs.existsSync(path.join(__dirname, '../.git'))) {
  test('git option', async function() {
    var file = path.join(__dirname, './fixture/simple.input.js');
    const result = await documentation.build([file], { github: true });
    normalize(result);
    expect(result).toMatchSnapshot();

    const md = await outputMarkdown(result, {});
    expect(md.toString()).toMatchSnapshot();
  });
}

test('document-exported error', async function() {
  var file = path.join(__dirname, 'fixture', 'document-exported-bad', 'x.js');
  try {
    await documentation.build([file], { documentExported: true });
  } catch (err) {
    expect(err.message.match(/Unable to find the value x/g)).toBeTruthy();
  }
});

test('external modules option', async function() {
  const result = await documentation.build(
    [path.join(__dirname, 'fixture', 'external.input.js')],
    {
      external: '(external|external/node_modules/*)'
    }
  );
  normalize(result);
  var outputfile = path.join(
    __dirname,
    'fixture',
    '_external-deps-included.json'
  );
  expect(result).toMatchSnapshot();
});

test('bad input', function() {
  glob
    .sync(path.join(__dirname, 'fixture/bad', '*.input.js'))
    .forEach(function(file) {
      test(path.basename(file), function() {
        return documentation
          .build([file], readOptionsFromFile(file))
          .then(res => {
            expect(res).toBe(undefined);
          })
          .catch(error => {
            // make error a serializable object
            error = JSON.parse(JSON.stringify(error));
            // remove system-specific path
            delete error.filename;
            delete error.codeFrame;
            expect(error).toMatchSnapshot();
          });
      });
    });
});

describe('html', function() {
  glob
    .sync(path.join(__dirname, 'fixture/html', '*.input.js'))
    .forEach(function(file) {
      test(path.basename(file), async function() {
        const result = await documentation.build(
          [file],
          readOptionsFromFile(file)
        );
        const html = await outputHtml(result, {});
        var clean = html
          .sort((a, b) => a.path > b.path)
          .filter(r => r.path.match(/(html)$/))
          .map(r => r.contents)
          .join('\n');
        expect(clean).toMatchSnapshot();
      });
    });
});

describe('outputs', function() {
  glob
    .sync(path.join(__dirname, 'fixture', '*.input.js'))
    .forEach(function(file) {
      test(path.basename(file), async function() {
        const result = await documentation.build(
          [file],
          readOptionsFromFile(file)
        );
        test('markdown', async function() {
          const md = await outputMarkdown(_.cloneDeep(result), {
            markdownToc: true
          });
          expect(result.toString()).toMatchSnapshot();
        });

        if (file.match(/es6.input.js/)) {
          test('no markdown TOC', async function() {
            const txt = await outputMarkdown(_.cloneDeep(result), {
              markdownToc: false
            });
            expect(result.toString()).toMatchSnapshot();
          });
        }

        test('markdown AST', async function() {
          const ast = await outputMarkdownAST(_.cloneDeep(result), {});
          expect(ast).toMatchSnapshot();
        });

        test('JSON', function() {
          normalize(result);
          result.forEach(function(comment) {
            validate(
              comment,
              documentationSchema.jsonSchema
            ).errors.forEach(function(error) {
              expect(error).toBeFalsy();
            });
          });
          expect(makePOJO(result)).toMatchSnapshot();
        });
      });
    });
});

test('highlightAuto md output', async function() {
  var file = path.join(
      __dirname,
      'fixture/auto_lang_hljs/multilanguage.input.js'
    ),
    hljsConfig = {
      hljs: { highlightAuto: true, languages: ['js', 'css', 'html'] }
    };

  const result = await documentation.build(file, {});
  const md = await outputMarkdown(result, hljsConfig);
  expect(md.toString()).toMatchSnapshot();
});

test('config', async function() {
  var file = path.join(__dirname, 'fixture', 'class.input.js');
  var outputfile = path.join(__dirname, 'fixture', 'class.config.output.md');
  const out = await documentation.build([file], {
    config: path.join(__dirname, 'fixture', 'simple.config.yml')
  });
  const md = await outputMarkdown(out, {});
  expect(md).toMatchSnapshot();
});

test('config with nested sections', async function() {
  var file = path.join(__dirname, 'fixture', 'sections.input.js');
  const out = await documentation.build([file], {
    config: path.join(__dirname, 'fixture', 'sections.config.yml')
  });
  const md = await outputMarkdown(out, {});
  expect(md).toMatchSnapshot();
});

test('multi-file input', async function() {
  const result = await documentation.build(
    [
      path.join(__dirname, 'fixture', 'simple.input.js'),
      path.join(__dirname, 'fixture', 'simple-two.input.js')
    ],
    {}
  );
  normalize(result);
  expect(result).toMatchSnapshot();
});

test('accepts simple relative paths', async function() {
  await pify(chdir)(__dirname);
  const data = await documentation.build(
    '__tests__/fixture/simple.input.js',
    {}
  );
  expect(data.length).toBe(1);
});

test('.lint', async function() {
  await pify(chdir)(__dirname);
  const data = await documentation.lint(
    '__tests__/fixture/simple.input.js',
    {}
  );
  expect(data).toBe('');
});

test('.lint with bad input', async function() {
  await pify(chdir)(__dirname);
  try {
    await documentation.lint('__tests__/fixture/bad/syntax.input', {
      parseExtension: ['input']
    });
  } catch (err) {
    expect(err).toBeTruthy();
  }
});
