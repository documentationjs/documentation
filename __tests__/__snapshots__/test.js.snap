// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config 1`] = `
"<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

## MyClass

This is my class, a demo thing.

**Properties**

-   \`howMany\` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** how many things it contains

### getFoo

Get the number 42

**Parameters**

-   \`getIt\` **[boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** whether to get the number

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** forty-two

### getUndefined

Get undefined

Returns **[undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/undefined)** does not return anything.

## Hello

World

"
`;

exports[`external modules option 1`] = `
Array [
  Object {
    "augments": Array [],
    "context": Object {
      "loc": SourceLocation {
        "end": Position {
          "column": 1,
          "line": 10,
        },
        "start": Position {
          "column": 0,
          "line": 8,
        },
      },
    },
    "description": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Position {
                "end": Object {
                  "column": 9,
                  "line": 1,
                  "offset": 8,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "text",
              "value": "I am in ",
            },
            Object {
              "position": Position {
                "end": Object {
                  "column": 28,
                  "line": 1,
                  "offset": 27,
                },
                "indent": Array [],
                "start": Object {
                  "column": 9,
                  "line": 1,
                  "offset": 8,
                },
              },
              "type": "inlineCode",
              "value": "external.input.js",
            },
            Object {
              "position": Position {
                "end": Object {
                  "column": 29,
                  "line": 1,
                  "offset": 28,
                },
                "indent": Array [],
                "start": Object {
                  "column": 28,
                  "line": 1,
                  "offset": 27,
                },
              },
              "type": "text",
              "value": ".",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 29,
              "line": 1,
              "offset": 28,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "paragraph",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 29,
          "line": 1,
          "offset": 28,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "root",
    },
    "errors": Array [],
    "examples": Array [],
    "kind": "function",
    "loc": SourceLocation {
      "end": Position {
        "column": 3,
        "line": 7,
      },
      "start": Position {
        "column": 0,
        "line": 5,
      },
    },
    "members": Object {
      "events": Array [],
      "global": Array [],
      "inner": Array [],
      "instance": Array [],
      "static": Array [],
    },
    "name": "foo",
    "namespace": "foo",
    "params": Array [],
    "path": Array [
      Object {
        "kind": "function",
        "name": "foo",
      },
    ],
    "properties": Array [],
    "returns": Array [],
    "sees": Array [],
    "tags": Array [],
    "throws": Array [],
    "todos": Array [],
  },
  Object {
    "augments": Array [],
    "context": Object {
      "loc": SourceLocation {
        "end": Position {
          "column": 2,
          "line": 10,
        },
        "start": Position {
          "column": 0,
          "line": 7,
        },
      },
    },
    "description": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Position {
                "end": Object {
                  "column": 38,
                  "line": 1,
                  "offset": 37,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "text",
              "value": "This function returns the number one.",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 38,
              "line": 1,
              "offset": 37,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "paragraph",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 38,
          "line": 1,
          "offset": 37,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "root",
    },
    "errors": Array [],
    "examples": Array [],
    "kind": "function",
    "loc": SourceLocation {
      "end": Position {
        "column": 3,
        "line": 6,
      },
      "start": Position {
        "column": 0,
        "line": 3,
      },
    },
    "members": Object {
      "events": Array [],
      "global": Array [],
      "inner": Array [],
      "instance": Array [],
      "static": Array [],
    },
    "name": "main",
    "namespace": "main",
    "params": Array [],
    "path": Array [
      Object {
        "kind": "function",
        "name": "main",
      },
    ],
    "properties": Array [],
    "returns": Array [
      Object {
        "description": Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 10,
                      "line": 1,
                      "offset": 9,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 1,
                      "line": 1,
                      "offset": 0,
                    },
                  },
                  "type": "text",
                  "value": "numberone",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 10,
                  "line": 1,
                  "offset": 9,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 10,
              "line": 1,
              "offset": 9,
            },
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "root",
        },
        "title": "returns",
        "type": Object {
          "name": "Number",
          "type": "NameExpression",
        },
      },
    ],
    "sees": Array [],
    "tags": Array [
      Object {
        "description": "numberone",
        "lineNumber": 2,
        "title": "returns",
        "type": Object {
          "name": "Number",
          "type": "NameExpression",
        },
      },
    ],
    "throws": Array [],
    "todos": Array [],
  },
  Object {
    "augments": Array [],
    "context": Object {
      "loc": SourceLocation {
        "end": Position {
          "column": 2,
          "line": 8,
        },
        "start": Position {
          "column": 0,
          "line": 5,
        },
      },
    },
    "description": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Position {
                "end": Object {
                  "column": 38,
                  "line": 1,
                  "offset": 37,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "text",
              "value": "This function returns the number one.",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 38,
              "line": 1,
              "offset": 37,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "paragraph",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 38,
          "line": 1,
          "offset": 37,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "root",
    },
    "errors": Array [],
    "examples": Array [],
    "kind": "function",
    "loc": SourceLocation {
      "end": Position {
        "column": 3,
        "line": 4,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "members": Object {
      "events": Array [],
      "global": Array [],
      "inner": Array [],
      "instance": Array [],
      "static": Array [],
    },
    "name": "index",
    "namespace": "index",
    "params": Array [],
    "path": Array [
      Object {
        "kind": "function",
        "name": "index",
      },
    ],
    "properties": Array [],
    "returns": Array [
      Object {
        "description": Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 10,
                      "line": 1,
                      "offset": 9,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 1,
                      "line": 1,
                      "offset": 0,
                    },
                  },
                  "type": "text",
                  "value": "numberone",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 10,
                  "line": 1,
                  "offset": 9,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 10,
              "line": 1,
              "offset": 9,
            },
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "root",
        },
        "title": "returns",
        "type": Object {
          "name": "Number",
          "type": "NameExpression",
        },
      },
    ],
    "sees": Array [],
    "tags": Array [
      Object {
        "description": "numberone",
        "lineNumber": 2,
        "title": "return",
        "type": Object {
          "name": "Number",
          "type": "NameExpression",
        },
      },
    ],
    "throws": Array [],
    "todos": Array [],
  },
]
`;

exports[`git option 1`] = `
Array [
  Object {
    "augments": Array [],
    "context": Object {
      "github": "[github]",
      "loc": SourceLocation {
        "end": Position {
          "column": 2,
          "line": 8,
        },
        "start": Position {
          "column": 0,
          "line": 5,
        },
      },
    },
    "description": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Position {
                "end": Object {
                  "column": 38,
                  "line": 1,
                  "offset": 37,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "text",
              "value": "This function returns the number one.",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 38,
              "line": 1,
              "offset": 37,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "paragraph",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 38,
          "line": 1,
          "offset": 37,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "root",
    },
    "errors": Array [],
    "examples": Array [],
    "kind": "function",
    "loc": SourceLocation {
      "end": Position {
        "column": 3,
        "line": 4,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "members": Object {
      "events": Array [],
      "global": Array [],
      "inner": Array [],
      "instance": Array [],
      "static": Array [],
    },
    "name": "simple.input",
    "namespace": "simple.input",
    "params": Array [],
    "path": Array [
      Object {
        "kind": "function",
        "name": "simple.input",
      },
    ],
    "properties": Array [],
    "returns": Array [
      Object {
        "description": Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 10,
                      "line": 1,
                      "offset": 9,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 1,
                      "line": 1,
                      "offset": 0,
                    },
                  },
                  "type": "text",
                  "value": "numberone",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 10,
                  "line": 1,
                  "offset": 9,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 10,
              "line": 1,
              "offset": 9,
            },
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "root",
        },
        "title": "returns",
        "type": Object {
          "name": "number",
          "type": "NameExpression",
        },
      },
    ],
    "sees": Array [],
    "tags": Array [
      Object {
        "description": "numberone",
        "lineNumber": 2,
        "title": "returns",
        "type": Object {
          "name": "number",
          "type": "NameExpression",
        },
      },
    ],
    "throws": Array [],
    "todos": Array [],
  },
]
`;

exports[`git option 2`] = `
"<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

## simple.input

[undefined:5-8](<> \\"Source code on GitHub\\")

This function returns the number one.

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** numberone
"
`;

exports[`highlightAuto md output 1`] = `
"<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

## multilanguage.input

**Extends Foo, Bar**

This function returns the number one.

**Examples**

\`\`\`js
var myFoo = new Foo('[data-foo]');
myFoo.foo(42);
\`\`\`

\`\`\`html
<p data-foo>Data-Foo Element in the dom</p>
\`\`\`

\`\`\`css
[data-foo] {
   background-color: red;
}
\`\`\`

-   Throws **[Error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error)** if you give it something
-   Throws **[TypeError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError)** if you give it something else

Returns **[Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** numberone
"
`;

exports[`multi-file input 1`] = `
Array [
  Object {
    "augments": Array [],
    "context": Object {
      "loc": SourceLocation {
        "end": Position {
          "column": 1,
          "line": 13,
        },
        "start": Position {
          "column": 0,
          "line": 10,
        },
      },
    },
    "description": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Position {
                "end": Object {
                  "column": 43,
                  "line": 1,
                  "offset": 42,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "text",
              "value": "This function returns the number plus two.",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 43,
              "line": 1,
              "offset": 42,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "paragraph",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 43,
          "line": 1,
          "offset": 42,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "root",
    },
    "errors": Array [],
    "examples": Array [
      Object {
        "description": "var result = returnTwo(4);
// result is 6",
      },
    ],
    "kind": "function",
    "loc": SourceLocation {
      "end": Position {
        "column": 3,
        "line": 9,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "members": Object {
      "events": Array [],
      "global": Array [],
      "inner": Array [],
      "instance": Array [],
      "static": Array [],
    },
    "name": "returnTwo",
    "namespace": "returnTwo",
    "params": Array [
      Object {
        "description": Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 11,
                      "line": 1,
                      "offset": 10,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 1,
                      "line": 1,
                      "offset": 0,
                    },
                  },
                  "type": "text",
                  "value": "the number",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 11,
                  "line": 1,
                  "offset": 10,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "root",
        },
        "lineNumber": 3,
        "name": "a",
        "title": "param",
        "type": Object {
          "name": "Number",
          "type": "NameExpression",
        },
      },
    ],
    "path": Array [
      Object {
        "kind": "function",
        "name": "returnTwo",
      },
    ],
    "properties": Array [],
    "returns": Array [
      Object {
        "description": Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 10,
                      "line": 1,
                      "offset": 9,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 1,
                      "line": 1,
                      "offset": 0,
                    },
                  },
                  "type": "text",
                  "value": "numbertwo",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 10,
                  "line": 1,
                  "offset": 9,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 10,
              "line": 1,
              "offset": 9,
            },
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "root",
        },
        "title": "returns",
        "type": Object {
          "name": "Number",
          "type": "NameExpression",
        },
      },
    ],
    "sees": Array [],
    "tags": Array [
      Object {
        "description": "the number",
        "lineNumber": 3,
        "name": "a",
        "title": "param",
        "type": Object {
          "name": "Number",
          "type": "NameExpression",
        },
      },
      Object {
        "description": "numbertwo",
        "lineNumber": 4,
        "title": "returns",
        "type": Object {
          "name": "Number",
          "type": "NameExpression",
        },
      },
      Object {
        "description": "var result = returnTwo(4);
// result is 6",
        "lineNumber": 5,
        "title": "example",
      },
    ],
    "throws": Array [],
    "todos": Array [],
  },
  Object {
    "augments": Array [],
    "context": Object {
      "loc": SourceLocation {
        "end": Position {
          "column": 2,
          "line": 8,
        },
        "start": Position {
          "column": 0,
          "line": 5,
        },
      },
    },
    "description": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "position": Position {
                "end": Object {
                  "column": 38,
                  "line": 1,
                  "offset": 37,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "text",
              "value": "This function returns the number one.",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 38,
              "line": 1,
              "offset": 37,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "paragraph",
        },
      ],
      "position": Object {
        "end": Object {
          "column": 38,
          "line": 1,
          "offset": 37,
        },
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "root",
    },
    "errors": Array [],
    "examples": Array [],
    "kind": "function",
    "loc": SourceLocation {
      "end": Position {
        "column": 3,
        "line": 4,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "members": Object {
      "events": Array [],
      "global": Array [],
      "inner": Array [],
      "instance": Array [],
      "static": Array [],
    },
    "name": "simple.input",
    "namespace": "simple.input",
    "params": Array [],
    "path": Array [
      Object {
        "kind": "function",
        "name": "simple.input",
      },
    ],
    "properties": Array [],
    "returns": Array [
      Object {
        "description": Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 10,
                      "line": 1,
                      "offset": 9,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 1,
                      "line": 1,
                      "offset": 0,
                    },
                  },
                  "type": "text",
                  "value": "numberone",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 10,
                  "line": 1,
                  "offset": 9,
                },
                "indent": Array [],
                "start": Object {
                  "column": 1,
                  "line": 1,
                  "offset": 0,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Object {
            "end": Object {
              "column": 10,
              "line": 1,
              "offset": 9,
            },
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "root",
        },
        "title": "returns",
        "type": Object {
          "name": "number",
          "type": "NameExpression",
        },
      },
    ],
    "sees": Array [],
    "tags": Array [
      Object {
        "description": "numberone",
        "lineNumber": 2,
        "title": "returns",
        "type": Object {
          "name": "number",
          "type": "NameExpression",
        },
      },
    ],
    "throws": Array [],
    "todos": Array [],
  },
]
`;
